### DISCUSSION TITLE
Router push using "as" parameter is not available in app dir

### DISCUSSION THREAD

**Discussion by dawi** [vercel/next.js#49448] [App Router]
Hi together,

I am currently migrating a CRA to Next.js for obvious reasons.

The CRA used react-router-dom, where it was possible to attach state to navigation links.

In Next.js this could be solved like this:

```tsx
router.push(`login?from=${router.pathname}`, '/login');
```

However this is only available via pages router and not via app router.

The [app-router-migration documentation](https://nextjs.org/docs/app/building-your-application/upgrading/app-router-migration) doesn't really help in that regard. It only says

> the concept of as has been removed from the new router

So my questions are:

* How can I solve this using app router? Do I have to implement a workaround for this or am I missing something?
* Will pages directory be deprecated in the long term or will it exist "forever"? Because when I have to do the migration at some point I would rather do it now.

Thanks in advance :)

**Comment by jintak0401** [comment]
I also really want to know about this. I heard appDir is stable, so I am trying migration. However, there are pretty many disadvantages copared to the pages router...
